### PR TITLE
Fix `Object.defineProperty()`

### PR DESCRIPTION
**Pull request by HalidOdat**
It changes the following:
 - Fix panic if first argument is not supplied.
 - Fix panic if second argument is not supplied.
 - Fix bug when the second argument is not a object.
 - Implemented `DefinePropertyOrThrow()`

### CODE REVIEW

**Review by RageKnify — APPROVED** [review]

File: boa/src/builtins/object/mod.rs [hunk 1]
```diff
@@ -289,19 +289,23 @@ impl Object {
...
+                .unwrap_or(&Value::undefined())
+                .to_property_key(context)?;
+            let desc = args
+                .get(2)
+                .unwrap_or(&Value::undefined())
+                .to_property_descriptor(context)?;
+
+            object.define_property_or_throw(key, desc, context)?;
+
+            Ok(Value::undefined())
```
Comment (tofpie) [inline boa/src/builtins/object/mod.rs:305]: You could also fix this one: `Object.defineProperty` should return the object itself, and not `undefined` (see [spec][spec])

[spec]: https://tc39.es/ecma262/#sec-object.defineproperty

Comment (HalidOdat) [reply boa/src/builtins/object/mod.rs:305]: > You could also fix this one: `Object.defineProperty` should return the object itself, and not `undefined` (see [spec](https://tc39.es/ecma262/#sec-object.defineproperty))

Nice catch. You are right, I'll fix it now :)

**Comment by RageKnify** [comment]
Somehow 680579a changed 2 test262 tests from passing to panicking, the change can be seen by looking at the edit history of the comment showing the results, https://github.com/boa-dev/boa/pull/1005#issuecomment-751942679

Don't understand how that change could introduce panicking.

**Comment by HalidOdat** [comment]
> Somehow [680579a](https://github.com/boa-dev/boa/commit/680579a9b6e818e0d135a353c106635693cf3a6c) changed 2 test262 tests from passing to panicking, the change can be seen by looking at the edit history of the comment showing the results, [#1005 (comment)](https://github.com/boa-dev/boa/pull/1005#issuecomment-751942679)
> 
> Don't understand how that change could introduce panicking.

It is probably be due to another unrelated bug. `Object.defineProperty` should return the object.

**Comment by tofpie** [comment]
The new panic is caused by the test [15.2.3.6-4-187][panic-test].

This test calls `Object.defineProperty` on the property `length` of an `Array` setting `writable` to false but without defining a value.

As `defineProperty` now returns the object, it is displayed, and the panic occurs here, as `length` is `undefined`:
https://github.com/boa-dev/boa/blob/67ee6faa5819ba5e97ae826244a88ce370e32070/boa/src/value/display.rs#L98-L109

It seems that setting only `writable` in a property should not change its value, so the `length` should remain a number.

[panic-test]: https://github.com/tc39/test262/blob/42bf3a9f7aed3790dc1d829f290dee033df41c5b/test/built-ins/Object/defineProperty/15.2.3.6-4-187.js#L1

**Pull request merged**